devito/arguments.py

```diff
@@ -1,152 +1,139 @@
-            the Dimension can come up with.
-            In the above two cases, the value passed in case 2 is "optional", since
-            a dimension "belongs" to multiple Function objects and each object might
-            lead to a different best-guess. During normal behaviour, verify is expected
-            to be called multiple times with different values and this method will
-            combine all the values received in the different invocations in a configurable
-            (self.reducer) way. However, if the user has explicitly passed in a value
-            (case 1 above), that value has to be respected and this best-guess behaviour
-            described above needs to be switched off.
-            Another thing to note is that, once a ScalarArgument has been passed a value
-            with enforce=True, any further calls to verify should not change the stored
-            value. The self._frozen flag is used to ensure that.
-            :param enforce: Turn off the best-guess behaviour and just use the value
-                            passed, ignoring any previously stored values.
-        """
-        if value is not None:
-            if self._value is not None and not enforce:
-                # We already have a value and the value passed
-                # here is not enforced (i.e. optional)
-                # If we aren't frozen, use the reducer
-                if not self._frozen:
-                    self._value = self.reducer(self._value, value)
-            else:
-                # Either this is the first time we're getting a value (self.value = None)
-                # or this is an enforced value
-                if enforce or not self._frozen:
-                    # If we're frozen, ignore, else use this value directly
-                    self._value = value
-                    if enforce:
-                        # If we were forced to use this value, make sure we don't change
-                        # it in the same invocation
-                        self._frozen = True
-        return self._value is not None
-
 
 class TensorArgument(Argument):
+
     """ Class representing tensor arguments that a kernel might expect.
         Most commonly used to pass numpy-like multi-dimensional arrays.
     """
-
     is_TensorArgument = True
 
-    def __init__(self, name, provider):
-        super(TensorArgument, self).__init__(name, provider, provider)
-
-    def verify(self, value):
-        if value is None:
-            value = self._value
-
-        verify = len(self.provider.shape) == len(value.shape)
-
-        verify = verify and all(d.verify(v) for d, v in
-                                zip(self.provider.indices, value.shape))
-        if verify:
-            self._value = value
-
-        return self._value is not None and verify
+    def __init__(self, provider, dependencies=None):
+        if dependencies is None:
+            dependencies = []
```

Comment:
pythonic way:
``dependences = dependences or []``

Reply:
I disagree: https://stackoverflow.com/questions/3901144/difference-between-if-x-and-if-x-is-not-none

Reply:
dependencies can be either ``None`` or something that evaluates to True, hence the pythonic way is preferable. I'm 99% sure this is common practice when using ``None`` as default argument value (instead of [], for example)

Reply:
It is unclear to me why you would want me to change a tighter test to one that is more prone to unintentional effects.

Reply:
it was nitpicking, we can ignore this